### PR TITLE
fix(a11y): ensure tooltip-labelled content is accessible

### PR DESCRIPTION
**Pull request by tille**
The problem is the tooltip element is not in the DOM unless focused, so use cases like icons that are labeled visually with tooltips don’t have proper screenreader representation because they are defacto invisible in the screen reader tree

so we want to auto-apply the aria-label whenever there is a tooltip, unless the user has already applied an aria label to the target element, in which case the describedby relationship is applied - this is for the use case where you have an icon with obvious visual meaning + a manual aria label as a substitute, and a longer explanatory text in the tooltip, which only needs to show on focus in either the visual or screen reader mode

what we found is that a whole bunch of places where we placed tooltips on icons had to have manual aria-labels added that were equal to the tooltip text, and then the current implementation would also add the describedby relation on focus, which was redundant because the text was the same (this strategy takes care of this situation)

### CODE REVIEW

File: src/components/tooltip/Tooltip.tsx [hunk 1]
```diff
@@ -219,6 +219,12 @@ class Tooltip extends React.Component<TooltipProps, State> {
             theme,
         } = this.props;
 
+        const tooltipText =
+            text && typeof text === 'object' ? (text as React.ReactElement).props.defaultMessage : String(text);
```
Comment (laughnan) [inline src/components/tooltip/Tooltip.tsx:223]: Instead of accessing `props.defaultMessage` (assuming this test is a `react-intl` MessageDescriptor) you can use `intl.formatMessage` to render the message as a string.

File: src/components/tooltip/Tooltip.tsx [hunk 2]
```diff
@@ -219,6 +219,12 @@ class Tooltip extends React.Component<TooltipProps, State> {
...
+
+        const ariaLabel = (React.Children.only(children) as React.ReactElement).props['aria-label'];
+        const labelInName = ariaLabel && tooltipText !== ariaLabel;
```
Comment (laughnan) [inline src/components/tooltip/Tooltip.tsx:226]: `[nit]` maybe making the variable `isLabelMatchingTooltipText`

File: src/components/tooltip/Tooltip.tsx [hunk 3]
```diff
@@ -219,6 +219,12 @@ class Tooltip extends React.Component<TooltipProps, State> {
             theme,
         } = this.props;
 
+        const tooltipText =
+            text && typeof text === 'object' ? (text as React.ReactElement).props.defaultMessage : String(text);
+
+        const ariaLabel = (React.Children.only(children) as React.ReactElement).props['aria-label'];
```
Comment (alexkrolick) [inline src/components/tooltip/Tooltip.tsx:225]: Rather than casting this, can you make sure `.props` is defined, so there isn't the possibility runtime error if the tooltip is wrapped around a string or other non-Element node?

File: src/components/tooltip/Tooltip.tsx (hunk 2 above)
Comment (laughnan) [reply src/components/tooltip/Tooltip.tsx:226]: +1

File: src/components/tooltip/Tooltip.tsx [hunk 4]
```diff
@@ -217,10 +218,17 @@ class Tooltip extends React.Component<TooltipProps, State> {
             stopBubble,
             text,
             theme,
+            intl,
         } = this.props;
 
-        const tooltipText =
-            text && typeof text === 'object' ? (text as React.ReactElement).props.defaultMessage : String(text);
+        const textElement = text as React.ReactElement;
```
Comment (alexkrolick) [inline src/components/tooltip/Tooltip.tsx:224]: ReactElement is not the same as ReactNode, so this cast is masking the real possible types for the `text` prop

File: src/components/tooltip/Tooltip.tsx [hunk 5]
```diff
@@ -217,8 +219,24 @@ class Tooltip extends React.Component<TooltipProps, State> {
...
+            intl,
         } = this.props;
 
+        let ariaLabel;
+        let tooltipText;
+
+        if (text) {
+            if (typeof text === 'string') {
+                tooltipText = text;
+            } else if (React.isValidElement(text) && getProp(text, 'props.id')) {
```
Comment (alexkrolick) [inline src/components/tooltip/Tooltip.tsx:231]: Can you check for the element type being FormattedMessage (because anything can have an `id`), and pull out the exact react-intl props here and when you pass them into formatMessage

```js
const { id, defaultMessage, description, values } = props;
intl.formatMessage({ id, defaultMessage, description }, values)
```

Comment (alexkrolick) [reply src/components/tooltip/Tooltip.tsx:231]: cc @megansmith-box @DanDeMicco 👀 can you verify if this approach conflicts with any of the ways product adoption tooltips work? Do we put rich text formatting in there?

Comment (tille) [reply src/components/tooltip/Tooltip.tsx:231]: @alexkrolick do you mean something like this?
```
      if (text) {
          if (typeof text === 'string') {
              tooltipText = text;
          } else if (React.isValidElement(text) && getProp(text, 'props')) {
              const { id, defaultMessage, description, values } = getProp(text, 'props');
              if (id && defaultMessage && description) {
                  tooltipText = intl.formatMessage({ id, defaultMessage, description }, values)
              }
          }
      }
```

Comment (alexkrolick) [reply src/components/tooltip/Tooltip.tsx:231]: Yes, like that

File: src/components/tooltip/Tooltip.tsx [hunk 6]
```diff
@@ -217,8 +219,27 @@ class Tooltip extends React.Component<TooltipProps, State> {
...
+                tooltipText = text;
+            } else if (React.isValidElement(text) && getProp(text, 'props')) {
+                const { id, defaultMessage, description, values } = getProp(text, 'props');
+                if (id && defaultMessage) {
+                    tooltipText = intl.formatMessage({ id, defaultMessage, description }, values);
+                }
+            }
+        }
+        if (getProp(children, 'props')) {
+            ariaLabel = getProp(children, 'props.aria-label');
```
Comment (megansmith-box) [inline src/components/tooltip/Tooltip.tsx:239]: will aria-label always be a prop on children? should we have a default in case it's not or does getProp return a default?

File: src/components/tooltip/Tooltip.tsx [hunk 7]
```diff
@@ -217,8 +219,27 @@ class Tooltip extends React.Component<TooltipProps, State> {
...
+            intl,
         } = this.props;
 
+        let ariaLabel;
+        let tooltipText;
+
+        if (text) {
+            if (typeof text === 'string') {
+                tooltipText = text;
+            } else if (React.isValidElement(text) && getProp(text, 'props')) {
```
Comment (DanDeMicco) [inline src/components/tooltip/Tooltip.tsx:231]: What is expected behavior for tooltips with not just a single element passed in? e.g. ones with a title and a body?

`GuideTooltip` is an example of this, but there are multiple variations of this in the codebase.

File: src/components/tooltip/Tooltip.tsx [hunk 8]
```diff
@@ -217,8 +219,27 @@ class Tooltip extends React.Component<TooltipProps, State> {
             stopBubble,
             text,
             theme,
+            intl,
         } = this.props;
 
+        let ariaLabel;
```
Comment (megansmith-box) [inline src/components/tooltip/Tooltip.tsx:225]: is undefined a valid ariaLabel?

Comment (tille) [reply src/components/tooltip/Tooltip.tsx:225]: yes it is a valid one, because if is undefined then it will be captured by the line
https://github.com/box/box-ui-elements/pull/2544/files#diff-ecdd760eb0b8901c98cc10e0b0266af590216f4f4619845564b2755d1ef90c55R241

File: src/components/tooltip/Tooltip.tsx (hunk 6 above)
Comment (tille) [reply src/components/tooltip/Tooltip.tsx:239]: yes exactly

File: src/components/tooltip/Tooltip.tsx (hunk 7 above)
Comment (tille) [reply src/components/tooltip/Tooltip.tsx:231]: Pretty God question

File: src/components/tooltip/Tooltip.tsx (hunk 8 above)
Comment (laughnan) [reply src/components/tooltip/Tooltip.tsx:225]: Is there a benefit in renaming `ariaLabel` to `childComponentAriaLabel` to keep it clear that it's not the tooltip's `aria-label`?

File: src/components/tooltip/Tooltip.tsx (hunk 7 above)
Comment (laughnan) [reply src/components/tooltip/Tooltip.tsx:231]: This logic should work fine for `CheckboxTooltip`  and `ImageTooltip`. The `GuideTooltip` is tricky because it builds a pretty complex child component in the `text` prop. It might make sense to address this in a follow-up. Theoretically, the logic above shouldn't impact (good or bad) it.

File: src/components/tooltip/Tooltip.tsx [hunk 9]
```diff
@@ -217,8 +219,27 @@ class Tooltip extends React.Component<TooltipProps, State> {
...
+        let tooltipText;
+
+        if (text) {
+            if (typeof text === 'string') {
+                tooltipText = text;
+            } else if (React.isValidElement(text) && getProp(text, 'props')) {
+                const { id, defaultMessage, description, values } = getProp(text, 'props');
```
Comment (laughnan) [inline src/components/tooltip/Tooltip.tsx:232]: I think this should be fine for the first pass (still wrapping my head around updating the child's `aria-label` or adding in a `describedby` label. Can we add comments in this conditional to explain that we're specifically looking for these props since they indicate a `FormattedMessage` component is passed into `text` so that it's clear?

File: src/components/tooltip/Tooltip.tsx (hunk 7 above)
Comment (tille) [reply src/components/tooltip/Tooltip.tsx:231]: Where can I find a real example of those kind of Tooltips? (ex. GuideTooltip)

```
For those Tooltips I guess it should have a `aria-describedBy` relation and `aria-hidden=false`
this strategy takes care of that (but Im needing a real example to test it)
```

File: src/components/tooltip/Tooltip.tsx (hunk 8 above)
Comment (tille) [reply src/components/tooltip/Tooltip.tsx:225]: yes!

File: src/components/tooltip/Tooltip.tsx (hunk 7 above)
Comment (laughnan) [reply src/components/tooltip/Tooltip.tsx:232]: Shared via Slack.

File: src/components/image-tooltip/__tests__/ImageTooltip.test.tsx [hunk 10]
```diff
@@ -1,17 +1,23 @@
...
 import { shallow, ShallowWrapper } from 'enzyme';
+import { MessageDescriptor } from 'react-intl';
 
 import Button from '../../button/Button';
 import ImageTooltip from '../ImageTooltip';
 // @ts-ignore flow import
 import testImageSrc from '../getTestImageSrc';
 
 describe('components/image-tooltip/ImageTooltip', () => {
+    const intlFake: any = {
```
Comment (laughnan) [inline src/components/image-tooltip/__tests__/ImageTooltip.test.tsx:11]: I'm seeing these linting errors:

```
/Users/alaughnan/Code/box-ui-elements/src/components/image-tooltip/ImageTooltip.stories.tsx
  11:17  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any

/Users/alaughnan/Code/box-ui-elements/src/components/image-tooltip/__tests__/ImageTooltip.test.tsx
  11:21  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any

/Users/alaughnan/Code/box-ui-elements/src/components/tooltip/__tests__/Tooltip.test.tsx
  13:21  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
```

**Review by laughnan — COMMENTED** [review]
I think to fix the issue in `GuideTooltip` you can change it's type definition slightly:

```
...
import Tooltip, { TooltipProps as BaseTooltipProps, TooltipTheme } from '../tooltip';
...
type TooltipProps = Omit<BaseTooltipProps, 'intl' | 'text' | 'theme'>;
```

Basically replacing `JSX.LibraryManagedAttributes<typeof Tooltip, Tooltip['props']>` with `BaseTooltipProps` (added to import statement). And I think you actually want type `TooltipProps = Omit<BaseTooltipProps, 'intl' | 'text' | 'theme'>;` (I added intl to the Omit)

**Comment by tille** [comment]
> I think to fix the issue in `GuideTooltip` you can change it's type definition slightly:
> 
> ```
> ...
> import Tooltip, { TooltipProps as BaseTooltipProps, TooltipTheme } from '../tooltip';
> ...
> type TooltipProps = Omit<BaseTooltipProps, 'intl' | 'text' | 'theme'>;
> ```
> 
> Basically replacing `JSX.LibraryManagedAttributes<typeof Tooltip, Tooltip['props']>` with `BaseTooltipProps` (added to import statement). And I think you actually want type `TooltipProps = Omit<BaseTooltipProps, 'intl' | 'text' | 'theme'>;` (I added intl to the Omit)

done!

File: src/components/image-tooltip/__tests__/ImageTooltip.test.tsx (hunk 10 above)
Comment (tille) [reply src/components/image-tooltip/__tests__/ImageTooltip.test.tsx:11]: those are warnings ^^

File: src/components/guide-tooltip/GuideTooltip.tsx [hunk 11]
```diff
@@ -2,15 +2,15 @@ import * as React from 'react';
 import { FormattedMessage } from 'react-intl';
 import classNames from 'classnames';
 
-import Tooltip, { TooltipTheme } from '../tooltip';
+// @ts-ignore flow import
```
Comment (laughnan) [inline src/components/guide-tooltip/GuideTooltip.tsx:5]: Do we need this?

Comment (tille) [reply src/components/guide-tooltip/GuideTooltip.tsx:5]: yes it needed for the cause `TooltipProps` is not passing ts linting
it has to be ignored in flow to pass all linting errors

File: src/features/collapsible-sidebar/__tests__/__snapshots__/CollapsibleSidebarLogo.test.js.snap [hunk 12]
```diff
@@ -246,67 +246,80 @@ exports[`components/core/collapsible-sidebar/CollapsibleSidebar render 2`] = `
...
           isDisabled={true}
           isTabbable={false}
           position="middle-right"
-          theme="default"
         >
-          <CollapsibleSidebarLogo__StyledToggleButton
-            aria-label="Expand"
-            className="bdl-CollapsibleSidebar-toggleButton"
-            type="button"
+          <Tooltip
```
Comment (laughnan) [inline src/features/collapsible-sidebar/__tests__/__snapshots__/CollapsibleSidebarLogo.test.js.snap:253]: Why does this render two `Tooltips`?

Comment (swfree) [reply src/features/collapsible-sidebar/__tests__/__snapshots__/CollapsibleSidebarLogo.test.js.snap:253]: This snapshot is generated with `mount` instead of `shallow`. Same thing happened with the [SidebarToggleButton snapshot test](https://github.com/box/box-ui-elements/blob/master/src/components/sidebar-toggle-button/__tests__/SidebarToggleButton.test.js#L7)

**Review by swfree — APPROVED** [review]

**Review by laughnan — CHANGES_REQUESTED** [review]
After internal discussions, here are the changes to get this merged:

- Revert larger pull request changes and just check `text` prop as a string (no longer checking FormattedMessage, etc)
- Don't add `aria-describedby` when the `aria-label` is the same

File: src/components/guide-tooltip/GuideTooltip.tsx (hunk 11 above)
Comment (alexkrolick) [reply src/components/guide-tooltip/GuideTooltip.tsx:5]: This file is in TS, and the import is in TS, so it shouldn't need an ignore

File: src/components/text-input/__tests__/__snapshots__/TextInput.test.js.snap [hunk 13]
```diff
@@ -14,9 +14,6 @@ exports[`components/text-input/TextInput should render text input with descripti
       some description
     </div>
     <Tooltip
-      constrainToScrollParent={false}
```
Comment (alexkrolick) [inline src/components/text-input/__tests__/__snapshots__/TextInput.test.js.snap:17]: What's causing these default props to be dropped from the snapshots?

Comment (tille) [reply src/components/text-input/__tests__/__snapshots__/TextInput.test.js.snap]: the InjectIntl wrapped, just remove it and updated snapshots

File: src/components/tooltip/__tests__/__snapshots__/Tooltip.test.tsx.snap [hunk 14]
```diff
@@ -3,7 +3,7 @@
 exports[`components/tooltip/Tooltip render() should match snapshot when stopBubble is set 1`] = `
 <div
   className="tooltip bdl-Tooltip"
-  id="tooltip19"
+  id="tooltip22"
```
Comment (jstoffan) [inline src/components/tooltip/__tests__/__snapshots__/Tooltip.test.tsx.snap:6]: Is this change needed or can we revert it?

File: src/components/tooltip/__tests__/Tooltip.test.tsx [hunk 15]
```diff
@@ -109,7 +109,7 @@ describe('components/tooltip/Tooltip', () => {
         test('should show tooltip when isShown state is true', () => {
             const wrapper = shallow(
                 <Tooltip text="hi">
-                    <button />
+                    <button aria-label="tales" />
```
Comment (jstoffan) [inline src/components/tooltip/__tests__/Tooltip.test.tsx:112]: Do we have a condition that tests a button with no aria label?

File: src/components/tooltip/Tooltip.tsx [hunk 16]
```diff
@@ -325,12 +340,18 @@ class Tooltip extends React.Component<TooltipProps, State> {
...
                 </div>
             </div>
         ) : (
-            <div className={classes} id={this.tooltipID} role="tooltip" aria-live="polite">
+            <div
+                className={classes}
+                id={this.tooltipID}
+                role="tooltip"
+                aria-live="polite"
+                aria-hidden={isLabelMatchingTooltipText}
```
[the remark this answers is not in the document]
Comment (tille) [reply src/components/tooltip/Tooltip.tsx:349]: this is to avoid repetition of texts being announced so if BOTH are the same means that `isLabelMatchingTooltipText` will be true and so aria-label and tooltipText doesn't have to be read with ONLY one of them being read is enough

File: src/components/tooltip/__tests__/__snapshots__/Tooltip.test.tsx.snap (hunk 14 above)
Comment (tille) [reply src/components/tooltip/__tests__/__snapshots__/Tooltip.test.tsx.snap:6]: don't know why its being updated but anyway it doesn't make any side effect 
the ID can be 19 or 22 and doesn't generate any problem

File: src/components/tooltip/Tooltip.tsx [hunk 17]
```diff
@@ -222,6 +223,9 @@ class Tooltip extends React.Component<TooltipProps, State> {
             theme,
         } = this.props;
 
+        const childAriaLabel = getProp(children, 'props.aria-label');
+        const isLabelMatchingTooltipText = Boolean(childAriaLabel && childAriaLabel === text);
```
Comment (jstoffan) [inline src/components/tooltip/Tooltip.tsx:227]: Can we go with `const isLabelMatchingTooltipText = !!childAriaLabel && childAriaLabel === text;`?

File: src/components/tooltip/Tooltip.tsx [hunk 18]
```diff
@@ -254,7 +258,10 @@ class Tooltip extends React.Component<TooltipProps, State> {
         }
 
         if (showTooltip) {
-            componentProps['aria-describedby'] = this.tooltipID;
+            if (!isLabelMatchingTooltipText || childAriaLabel === undefined) {
```
Comment (jstoffan) [inline src/components/tooltip/Tooltip.tsx:261]: Do we have any unit tests for this change?

Comment (tille) [reply src/components/tooltip/Tooltip.tsx:261]: yes `when aria-label does not exists describedBy relation should exists` 
https://github.com/box/box-ui-elements/blob/30f5e4d5d533be5a7de20804cc705d4ae818c5b8/src/components/tooltip/__tests__/Tooltip.test.tsx#L122

File: src/components/tooltip/Tooltip.tsx (hunk 17 above)
Comment (tille) [reply src/components/tooltip/Tooltip.tsx:227]: hmm I think... yes

File: src/components/tooltip/Tooltip.tsx (hunk 18 above)
Comment (jstoffan) [reply src/components/tooltip/Tooltip.tsx:261]: It looks like that's testing the positive case. Do we have any assertions for the negative case?

Comment (tille) [reply src/components/tooltip/Tooltip.tsx:261]: https://github.com/box/box-ui-elements/pull/2544/commits/31c7a92b5882470b4fd730fd35ffd6f8f5a28b36

**Review by jstoffan — DISMISSED** [review]
Looks good to me!

**Review by laughnan — APPROVED** [review]
🎉

**Review by jstoffan — APPROVED** [review]
